ghcide/src/Development/IDE/Plugin/ChangeTypeAction.hs

```diff
@@ -0,0 +1,130 @@
+                                                 TextDocumentIdentifier (TextDocumentIdentifier),
+                                                 TextEdit (TextEdit), Uri,
+                                                 WorkspaceEdit (WorkspaceEdit),
+                                                 type (|?) (InR))
+import           Text.Regex.TDFA                ((=~))
+
+descriptor :: PluginId -> PluginDescriptor IdeState
+descriptor plId = (defaultPluginDescriptor plId) { pluginHandlers = mkPluginHandler STextDocumentCodeAction codeActionHandler }
+
+codeActionHandler :: PluginMethodHandler IdeState 'TextDocumentCodeAction
+codeActionHandler ideState plId CodeActionParams {_textDocument = TextDocumentIdentifier uri, _context = CodeActionContext (List diags) _} = response $ do
+      nfp <- getNormalizedFilePath plId (TextDocumentIdentifier uri)
+      decls <- getDecls ideState nfp
+      let actions = generateActions uri decls diags
+      pure $ List actions
+
+getDecls :: MonadIO m => IdeState -> NormalizedFilePath -> ExceptT String m [LHsDecl GhcPs]
+getDecls state = handleMaybeM "Error: Could not get Parsed Module"
+    . liftIO
+    . fmap (fmap (hsmodDecls . unLoc . pm_parsed_source))
+    . runAction "changeSignature.GetParsedModule" state
+    . use GetParsedModule
+
+data ChangeSignature = ChangeSignature { expectedType :: Text
+                                       , actualType   :: Text
+                                       , declName     :: Text
+                                       , declSrcSpan  :: Maybe SrcSpan
+                                       , diagnostic   :: Diagnostic
+                                       , uri          :: Uri
+                                       }
+
+-- Needed to trackdown OccNames in signatures
+type SigName p = (HasOccName (IdP (GhcPass p)))
+
+-- | Generate CodeActions from a list of Diagnostics
+generateActions :: SigName p => Uri -> [LHsDecl (GhcPass p)] -> [Diagnostic] -> [Command |? CodeAction]
+generateActions uri = mapMaybe . generateAction uri
+
+-- | Create a CodeAction from a Diagnostic
+generateAction :: SigName p => Uri -> [LHsDecl (GhcPass p)] -> Diagnostic -> Maybe (Command |? CodeAction)
+generateAction uri decls diag = diagnosticToChangeSig uri decls diag >>= changeSigToCodeAction
+
+-- | Convert a diagnostic into a ChangeSignature and add the proper SrcSpan
+diagnosticToChangeSig :: SigName p => Uri -> [LHsDecl (GhcPass p)] -> Diagnostic -> Maybe ChangeSignature
+diagnosticToChangeSig uri decls diag = addSrcSpan decls <$> matchingDiagnostic uri diag
+
+-- | If a diagnostic has the proper message create a ChangeSignature from it
+matchingDiagnostic :: Uri -> Diagnostic -> Maybe ChangeSignature
+matchingDiagnostic uri diag@Diagnostic{_message} = unwrapMatch $ _message =~ expectedMessage
+    where
+        expectedMessage = "Expected type: (.+)\n +Actual type: (.+)\n.*\n +In an equation for ‘(.+)’" :: Text
```

Comment:
I wonder if it's worth adding a unit test or two just for the regex match with a couple of expected error messages? It seems like that's often the most fallible part of these plugins, so it might be nice to have something that asserts: yes, it really does match errors like this and that.

Obviously that doesn't help us if the error messages change, but we have the tests with real GHCs for that!

Reply:
Also, I thought this would do the wrong thing for some error messages, right? In the issue you pointed out that sometimes the error will highlight a *part* of the type that's wrong. We don't want to then go in and replace the whole signature in that case! Could do with some tests for the cases we shouldn't pick up and maybe a comment explaining why we think we're safe.

Reply:
Yeah, I hadn't dealt with most of the edge cases yet. I was just curious if someone much smarter than I knew of some mechanism of the GHC API that would be useful to look at. It seems like this is the only way until the new Diagnostics in `9.4`. 

And yes I have local unit tests just haven't bothered pushing them yet.